lib/maths/unittest/COutliersTest.cc

```diff
@@ -505,6 +505,80 @@ void COutliersTest::testFeatureInfluences() {
+    }};
+    TFactoryFunc toOnDiskDataFrame{[](const TPointVec& points) {
+        return test::CDataFrameTestUtils::toOnDiskDataFrame(
+            boost::filesystem::current_path().string(), points);
+    }};
+    TFactoryFunc factories[]{toMainMemoryDataFrame, toOnDiskDataFrame};
+
+    std::size_t numberPartitions[]{1, 3};
+    maths::COutliers::EMethod methods[]{maths::COutliers::E_Ensemble, maths::COutliers::E_Lof};
+    std::size_t numberNeighbours[]{0, 5};
+    bool computeFeatureInfluences[]{true, false};
+
+    std::size_t numberInliers{40000};
+    std::size_t numberOutliers{500};
+    std::size_t numberPoints{numberInliers + numberOutliers};
+
+    test::CRandomNumbers rng;
+
+    TPointVec points;
+    gaussianWithUniformNoise(rng, numberInliers, numberOutliers, points);
+
+    core::startDefaultAsyncExecutor(3);
+
+    for (std::size_t i = 0; i < 2; ++i) {
+
+        LOG_DEBUG(<< "# partitions = " << numberPartitions[i]);
+
+        auto frame = factories[i](points);
+
+        maths::COutliers::SComputeParameters params{2, // Number threads
+                                                    numberPartitions[i],
+                                                    true, // Standardize columns
+                                                    methods[i],
+                                                    numberNeighbours[i],
+                                                    computeFeatureInfluences[i],
+                                                    0.05}; // Outlier fraction
+
+        std::int64_t estimatedMemoryUsage(
+            core::CDataFrame::estimateMemoryUsage(i == 0, 40500, 6) +
+            maths::COutliers::estimateMemoryUsedByCompute(
+                params, numberPoints,
+                (numberPoints + numberPartitions[i] - 1) / numberPartitions[i],
+                6 /*dimension*/));
+
+        std::atomic<std::int64_t> memoryUsage{0};
+        std::atomic<std::int64_t> maxMemoryUsage{0};
+
+        maths::COutliers::compute(
+            params, *frame, [](double) {},
+            [&](std::int64_t delta) {
+                std::int64_t memoryUsage_{memoryUsage.fetch_add(delta)};
+
+                std::int64_t prevMaxMemoryUsage{maxMemoryUsage};
+                while (prevMaxMemoryUsage < memoryUsage_ &&
+                       maxMemoryUsage.compare_exchange_weak(prevMaxMemoryUsage,
+                                                            memoryUsage_) == false) {
+                }
+                LOG_TRACE(<< "current memory = " << memoryUsage_
+                          << ", high water mark " << maxMemoryUsage.load());
+            });
+
+        LOG_DEBUG(<< "estimated peak memory = " << estimatedMemoryUsage);
+        LOG_DEBUG(<< "high water mark = " << maxMemoryUsage);
+        CPPUNIT_ASSERT(std::fabs(maxMemoryUsage - estimatedMemoryUsage) <
+                       std::max(maxMemoryUsage.load(), estimatedMemoryUsage) / 3);
```

Comment:
I think this assertion is probably _only just_ passing on macOS.

On Linux and Windows it's failing as follows:

```
2019-02-28 11:18:37,472 Coordinated Universal Time [2840] DEBUG COutliersTest.cc@539 # partitions = 1
2019-02-28 11:18:50,340 Coordinated Universal Time [2840] DEBUG COutliersTest.cc@575 estimated peak memory = 18251480
2019-02-28 11:18:50,341 Coordinated Universal Time [2840] DEBUG COutliersTest.cc@576 high water mark = 16206964
2019-02-28 11:18:50,341 Coordinated Universal Time [2840] DEBUG COutliersTest.cc@539 # partitions = 3
2019-02-28 11:19:00,541 Coordinated Universal Time [2840] DEBUG COutliersTest.cc@575 estimated peak memory = 5482040
2019-02-28 11:19:00,541 Coordinated Universal Time [2840] DEBUG COutliersTest.cc@576 high water mark = 8721646

2019-02-28 11:16:55,854 GMT [3918] DEBUG COutliersTest.cc@539 # partitions = 1
2019-02-28 11:17:07,612 GMT [3918] DEBUG COutliersTest.cc@575 estimated peak memory = 18251480
2019-02-28 11:17:07,612 GMT [3918] DEBUG COutliersTest.cc@576 high water mark = 15882008
2019-02-28 11:17:07,612 GMT [3918] DEBUG COutliersTest.cc@539 # partitions = 3
2019-02-28 11:17:16,448 GMT [3918] DEBUG COutliersTest.cc@575 estimated peak memory = 5482040
2019-02-28 11:17:16,448 GMT [3918] DEBUG COutliersTest.cc@576 high water mark = 8720325
```

It seems like the estimate is out by very close to 1/3rd at the moment.  Can anything be done to make it better?  Or if not then we shouldn't assert 1/3rd.

Reply:
It should be possible to improve this bound and I think we should try and improve it. I haven't fully debugged the source of the discrepancy for partitioned analyses. I get consistently pretty good results for unpartitioned analyses. Also this particular case does some non-standard things (such as overriding a variety of parameters). I was thinking I'd follow up with work targeting the error in a separate PR. I propose relaxing this threshold for now, but leaving a TODO to improve the estimate.